File: src/lib/ax-size-order.ts
/**
 * Size chip ordering — small → large.
 * Handles Arc'teryx inseam (28-S), letter sizes, and compound labels
 * like YSL "XS / GB XS" or "F34 / GB 6".
 */

const LETTER_ORDER: Record<string, number> = {
  XXXS: 0,
  XXS: 1,
  XS: 2,
  S: 3,
  M: 4,
  L: 5,
  XL: 6,
  XXL: 7,
  "2XL": 7,
  XXXL: 8,
  "3XL": 8,
  "4XL": 9,
  "5XL": 10,
  OS: 11,
  "ONE SIZE": 11,
};

const LETTER_RE =
  /\b(XXXS|XXS|XS|S|M|L|XL|XXL|XXXL|2XL|3XL|4XL|5XL|OS|ONE\s*SIZE)\b/i;
const INSEAM_RE = /^(\d+(?:\.\d+)?)\s*[- ]?\s*([SRT])$/i;

function waistRank(raw: string): number {
  const s = raw.trim();
  if (s === "00") return -1;
  const n = Number(s);
  return Number.isFinite(n) ? n : 9999;
}

export function axSizeSortKey(size: string): [number, number, number, string] {
  const s = (size || "").trim();
  if (!s) return [9, 0, 0, ""];

  const inseam = s.match(INSEAM_RE);
  if (inseam) {
    const waist = inseam[1];
    const length = inseam[2].toUpperCase();
    return [0, "SRT".indexOf(length), waistRank(waist), s];
  }

  // Exact letter match (XS, M, …)
  const exact = LETTER_ORDER[s.toUpperCase()];
  if (exact !== undefined) return [1, exact, 0, s];

  // Compound: "XS / GB XS", "YSL S / GB S", "L / GB L"
  const letter = s.match(LETTER_RE);
  if (letter) {
    const tok = letter[1].toUpperCase().replace(/\s+/g, " ");
    const rank = LETTER_ORDER[tok === "ONE SIZE" ? "ONE SIZE" : tok];
    if (rank !== undefined) return [1, rank, 0, s];
  }

  // Fashion numeric: "F34 / GB 6", "IT 40", bare "34"
  const f = s.match(/\bF(\d{2})\b/i);
  if (f) return [2, Number(f[1]), 0, s];
  const gb = s.match(/\bGB\s*(\d{1,2})\b/i);
  if (gb) return [2, Number(gb[1]), 0, s];
  const bare = s.match(/^(\d{2})(?:\.\d+)?$/);
  if (bare) return [2, Number(bare[1]), 0, s];

  const n = Number(s);
  if (Number.isFinite(n)) return [2, n, 0, s];

  return [3, 0, 0, s.toLowerCase()];
}

export function compareAxSizes(a: string, b: string): number {
  const ka = axSizeSortKey(a);
  const kb = axSizeSortKey(b);
  for (let i = 0; i < 3; i++) {
    if (ka[i] !== kb[i]) return (ka[i] as number) - (kb[i] as number);
  }
  return ka[3] < kb[3] ? -1 : ka[3] > kb[3] ? 1 : 0;
}

export function isArcteryxProduct(product: {
  brand?: string;
  tags?: string[];
  id?: string;
}): boolean {
  if (product.brand === "아크테릭스" || product.brand === "Arc'teryx") return true;
  if (product.tags?.some((t) => /arcteryx|아크테릭스/i.test(t))) return true;
  if (product.id && /^(axa|axo|ax)-/i.test(product.id)) return true;
  return false;
}

File: scripts/ax_size_order.py
"""Arc'teryx size ordering — match official site (inseam S→R→T, then numeric)."""
from __future__ import annotations

import re

_LETTER_ORDER = {
    "XXS": 0,
    "XS": 1,
    "S": 2,
    "M": 3,
    "L": 4,
    "XL": 5,
    "XXL": 6,
    "2XL": 6,
    "XXXL": 7,
    "3XL": 7,
    "4XL": 8,
    "OS": 9,
    "ONE SIZE": 9,
}

# Pant / short waist + inseam: 00S, 0-R, 30 R, 32S, …
_INSEAM_RE = re.compile(
    r"^(\d+(?:\.\d+)?)\s*[- ]?\s*([SRT])$",
    re.IGNORECASE,
)


def _waist_rank(raw: str) -> float:
    """00 before 0 before 2 … (float('00')==float('0') so special-case)."""
    s = raw.strip()
    if s == "00":
        return -1.0
    try:
        return float(s)
    except ValueError:
        return 9999.0


def ax_size_sort_key(size: str) -> tuple:
    s = (size or "").strip()
    if not s:
        return (9, 0, 0, "")

    m = _INSEAM_RE.match(s)
    if m:
        waist, length = m.group(1), m.group(2).upper()
        return (0, "SRT".index(length), _waist_rank(waist), s)

    letter = _LETTER_ORDER.get(s.upper())
    if letter is not None:
        return (1, letter, 0, s)

    try:
        return (2, float(s), 0, s)
    except ValueError:
        return (3, 0, 0, s.lower())


def sort_ax_sizes(sizes: list[str]) -> list[str]:
    """Stable unique sort in Arc'teryx official display order."""
    seen: set[str] = set()
    uniq: list[str] = []
    for s in sizes:
        if s is None:
            continue
        t = str(s).strip()
        if not t or t in seen:
            continue
        seen.add(t)
        uniq.append(t)
    return sorted(uniq, key=ax_size_sort_key)
